File: windows/cpp/samples/DataBurnerCmd/Burner.cpp
#include "stdafx.h"
#include "Burner.h"
#include "DataStreamImpl.h"

Burner::Burner(void)
{
	m_pEngine = NULL;
	m_pEnumerator = NULL;
	m_pDevice = NULL;
	m_pDataDisc = NULL;

	m_isOpen = false;
}

Burner::~Burner(void)
{
	Close();
}

bool Burner::Open()
{	
	if (m_isOpen)
		return true;

	// Enable trace log
	Library::enableTraceLog(NULL, TRUE);

	m_pEngine = Library::createEngine();
	if (!m_pEngine->initialize()) 
	{
		throw BurnerException(m_pEngine);
	}

	m_isOpen = true;
	return true;
}

void Burner::Close()
{
	CleanupDataDisc();
	CleanupDevice();
	CleanupDeviceEnum();
	CleanupEngine();

	Library::disableTraceLog();

	m_isOpen = false;
}
void Burner::SelectDevice(int32_t deviceIndex, bool exclusive)
{
	CleanupDeviceEnum();
	CleanupDevice();
	m_pEnumerator = m_pEngine->createDeviceEnumerator();

	m_pDevice = m_pEnumerator->createDevice(deviceIndex, exclusive ? 1 : 0);

	if (NULL == m_pDevice)
	{
		throw BurnerException(m_pEnumerator);
	}
}
const DeviceVector& Burner::EnumerateDevices()
{
	if (!m_isOpen)
	{
		throw BurnerException(BE_BURNER_NOT_OPEN, BE_BURNER_NOT_OPEN_TEXT);
	}

	m_DeviceVector.clear();

	CleanupDeviceEnum();
	m_pEnumerator = m_pEngine->createDeviceEnumerator();

	int32_t nDevices = m_pEnumerator->count();
	if (0 == nDevices) 
	{
		throw BurnerException(BE_NO_DEVICES, BE_NO_DEVICES_TEXT);
	}

	for (int32_t i = 0; i < nDevices; i++) 
	{
		Device* pDevice = m_pEnumerator->createDevice(i);
		if (NULL != pDevice)
		{
			DeviceInfo dev;
			dev.Index = i;
			dev.Title = get_DeviceTitle(pDevice);
			dev.DriveLetter = pDevice->driveLetter();

			m_DeviceVector.push_back(dev);

			pDevice->release();
		}
		else
		{
			throw BurnerException(m_pEnumerator);
		}
	}

	return m_DeviceVector;
}

void Burner::set_Callback(IBurnerCallback* value)
{
	m_pCallback = value;
}
void Burner::BurnImage(const ImageBurnSettings& settings)
{
	if (NULL != m_pDevice)
	{
		PrepareMedium();

		CleanupDataDisc();
		m_pDataDisc = Library::createDataDisc();

		// Tell DataDisc which device it should use. Here we just pass the device the user selected.
		m_pDataDisc->setDevice(m_pDevice);

		// Set the write method
		if (m_pDevice->isMediaBD())
		{
			m_pDataDisc->setWriteMethod(WriteMethod::BluRay);
		}
		else if (m_pDevice->isMediaDVD())
		{
			// The best for DVD Media is DVD Incremental
			m_pDataDisc->setWriteMethod(WriteMethod::DVDIncremental);
		}
		else
		{
			// The best for CD Media is Track-At-Once (TAO). 
			m_pDataDisc->setWriteMethod(WriteMethod::Tao);
		}

		// Perform a real burn
		m_pDataDisc->setSimulateBurn(FALSE);

		// Set a callback object for the burning progress
		m_pDataDisc->setCallback(this);

		// Burn it
		CFileStream dataStream(settings.get_ImageFile().c_str());
		if (!m_pDataDisc->writeImageToDiscEx(&dataStream, NULL))
		{
			throw BurnerException(m_pDataDisc);
		}
	}
	else
	{
		throw BurnerException(BE_DEVICE_NOT_SET, BE_DEVICE_NOT_SET_TEXT);
	}
}

void Burner::BurnPacket(const PacketBurnSettings& settings)
{
	if (NULL != m_pDevice)
	{
		PrepareMedium();

		// Get next writable address
		int32_t lStartAddress = GetStartAddress(m_pDevice);
		if (-1 == lStartAddress)
		{
			throw BurnerException(m_pDevice);
		}

		// indicates whether we should load layout from the last track or not.
		BOOL fLoadLastTrack  = PBO_APPEND == settings.get_Option() || PBO_FINALIZE == settings.get_Option();

		CleanupDataDisc();
		// Create DataDisc object 
		m_pDataDisc = Library::createDataDisc();
		
		// Set device 
		m_pDataDisc->setDevice(m_pDevice);

		// Set other parameters
		SetPacketParameters(m_pDataDisc, settings);

		// Get the last track number from the last session if multi-session option was specified
		int nPrevTrackNumber = fLoadLastTrack ? GetLastTrackNumber(m_pDevice, false) : 0;
		m_pDataDisc->setLayoutLoadTrack(nPrevTrackNumber);

		// Set write speed
		int iMaxWriteSpeedKB = m_pDevice->maxWriteSpeedKB();
		//_tprintf(_T("Setting write speed to: %.02f\n"), (double)iMaxWriteSpeedKB / (pDevice->GetMediaIsDVD() ? SPEED_DVD_1X_KB : SPEED_CD_1X_KB));
		m_pDevice->setWriteSpeedKB(iMaxWriteSpeedKB);

		// Set callback
		m_pDataDisc->setCallback(this);

		// Set the session start address. Must do this before intializing the directory structure.
		m_pDataDisc->setSessionStartAddress(lStartAddress);

		// Set image layout
		BOOL bRes = m_pDataDisc->setImageLayoutFromFolder(settings.get_FolderSrc().c_str());
		if (!bRes) 
		{
			throw BurnerException(m_pDataDisc);
		}

		// Burn 
		while (true)
		{
			// Try to write the image
			bRes = m_pDataDisc->writeToDisc();
			if (!bRes)
			{
				// Check if the error is: Cannot load image layout.
				// If so most likely it is an empty formatted DVD+RW or empty formatted DVD-RW RO with one track. 
				if((m_pDataDisc->error()->facility() == ErrorFacility::DataDisc) &&
					(m_pDataDisc->error()->code() == DataDiscError::CannotLoadImageLayout))
				{
					// Set to 0 to disable previous data session loading
					m_pDataDisc->setLayoutLoadTrack(0);

					// try to write it again
					continue;
				}
			}

			break;
		}

		if (!bRes) 
		{
			throw BurnerException(m_pDataDisc);
		}
	}
	else
	{
		throw BurnerException(BE_DEVICE_NOT_SET, BE_DEVICE_NOT_SET_TEXT);
	}
}



void Burner::BurnSimple(const SimpleBurnSettings& settings)
{
	if (NULL != m_pDevice)
	{
		PrepareMedium();

		// Clean previous recordings
		if (SBO_MERGE != settings.get_Option())
			Clean();

		// Tell DataDisc which device it should use. Here we just pass the device the user selected.
		CleanupDataDisc();
		m_pDataDisc = Library::createDataDisc();
		m_pDataDisc->setDevice(m_pDevice);

		// Set data image volume label
		SetVolumeProperties(m_pDataDisc, settings.get_VolumeLabel(), settings.get_ImageType());

		// Set image type. Must be done before calling SetImageLayoutFromFolder.
		m_pDataDisc->setImageType(settings.get_ImageType());

		// Set the session start address. Must do this before intializing the directory structure.
		m_pDataDisc->setSessionStartAddress(m_pDevice->newSessionStartAddress());

		// Load last complete track multi-session
		if (SBO_MERGE == settings.get_Option())
		{
			int nLastCompleteTrack = GetLastTrackNumber(m_pDevice, true);
			m_pDataDisc->setLayoutLoadTrack(nLastCompleteTrack);
		}

		// The easiest is to use the SetImageLayoutFromFolder method to specify a folder the content of which should be recorded to the disc.
		if (!m_pDataDisc->setImageLayoutFromFolder(settings.get_FolderSrc().c_str()))
		{
			throw BurnerException(m_pDataDisc);
		}

		// Set the write method
		if (m_pDevice->isMediaBD())
		{
			m_pDataDisc->setWriteMethod(WriteMethod::BluRay);
		}
		else if (m_pDevice->isMediaDVD())
		{
			// The best for DVD Media is DVD Incremental
			m_pDataDisc->setWriteMethod(WriteMethod::DVDIncremental);
		}
		else
		{
			// The best for CD Media is Track-At-Once (TAO). 
			m_pDataDisc->setWriteMethod(WriteMethod::Tao);
		}

		// Perform a real burn
		m_pDataDisc->setSimulateBurn(FALSE);

		// Set a callback object for the burning progress
		m_pDataDisc->setCallback(this);

		// Do not close disc to allow multisession
		m_pDataDisc->setCloseDisc(FALSE);

		// Set write speed
		int iMaxWriteSpeedKB = m_pDevice->maxWriteSpeedKB();
		m_pDevice->setWriteSpeedKB(iMaxWriteSpeedKB);

		// Burn it
		if (!m_pDataDisc->writeToDisc())
		{
			throw BurnerException(m_pDataDisc);
		}
	}
	else
	{
		throw BurnerException(BE_DEVICE_NOT_SET, BE_DEVICE_NOT_SET_TEXT);
	}
}
void Burner::Erase()
{
	if (NULL != m_pDevice)
	{
		m_pDevice->setCallback(this);
		bool bCleanResult = true;
		MediaProfile::Enum mp = m_pDevice->mediaProfile();
		switch(mp)
		{
			// DVD-RW, Sequential Recording (default for new DVD-RW)
			case MediaProfile::DVDMinusRWSeq:
				bCleanResult = 1 == m_pDevice->erase(EraseType::Minimal);
			break;

			case MediaProfile::CDRW:
				bCleanResult = 1 == m_pDevice->erase(EraseType::Minimal);
			break;

		}
		if (!bCleanResult)
		{
			throw BurnerException(m_pDevice);
		}
	}
	else
	{
		throw BurnerException(BE_DEVICE_NOT_SET, BE_DEVICE_NOT_SET_TEXT);
	}
}


void Burner::Format()
{
	if (NULL != m_pDevice)
	{
		m_pDevice->setCallback(this);
		bool bCleanResult = true;
		MediaProfile::Enum mp = m_pDevice->mediaProfile();
		switch(mp)
		{
			// DVD+RW (needs to be formatted before the disc can be used)
			case MediaProfile::DVDPlusRW:
				bCleanResult = FormatDVDPlusRW();
			break;

			// DVD-RW, Restricted Overwrite (DVD-RW was formatted initially)
			case MediaProfile::DVDMinusRWRO:
				bCleanResult = 1 == m_pDevice->format(FormatType::DVDMinusRWQuick);
			break;

			// BD-RE
			case MediaProfile::BDRE:
				bCleanResult = 1 == m_pDevice->formatBD(BDFormatType::BDFull, BDFormatSubType::BDREQuickReformat);
			break;

			// Format for Pseudo-Overwrite (POW)
			case MediaProfile::BDRSrm:
				bCleanResult = 1 == m_pDevice->formatBD(BDFormatType::BDFull, BDFormatSubType::BDRSrmPow);
			break;

		}
		if (!bCleanResult)
		{
			throw BurnerException(m_pDevice);
		}
	}
	else
	{
		throw BurnerException(BE_DEVICE_NOT_SET, BE_DEVICE_NOT_SET_TEXT);
	}
}

bool Burner::FormatDVDPlusRW()
{
	bool bCleanResult = true;
	if (NULL != m_pDevice)
	{
		BgFormatStatus::Enum fmt = m_pDevice->bgFormatStatus();
		switch(fmt)
		{
			case BgFormatStatus::NotFormatted:
				bCleanResult = 1 == m_pDevice->format(FormatType::DVDPlusRWFull);
				break;
			case BgFormatStatus::Partial:
				bCleanResult = 1 == m_pDevice->format(FormatType::DVDPlusRWRestart);
				break;
		}
	}
	else
	{
		throw BurnerException(BE_DEVICE_NOT_SET, BE_DEVICE_NOT_SET_TEXT);
	}

	return bCleanResult;
}

void Burner::PrepareMedium()
{
	if (NULL != m_pDevice)
	{
		MediaProfile::Enum mp = m_pDevice->mediaProfile();
		switch(mp)
		{
			// DVD+RW (needs to be formatted before the disc can be used)
			case MediaProfile::DVDPlusRW:
				m_pDevice->setCallback(this);
				FormatDVDPlusRW();
				break;
		}
	}
	else
	{
		throw BurnerException(BE_DEVICE_NOT_SET, BE_DEVICE_NOT_SET_TEXT);
	}
}

void Burner::Clean()
{
	if (NULL != m_pDevice)
	{
		MediaProfile::Enum mp = m_pDevice->mediaProfile();
		switch(mp)
		{
			case MediaProfile::DVDPlusRW:
			case MediaProfile::DVDMinusRWRO:
			case MediaProfile::BDRE:
			case MediaProfile::BDRSrm:
				Format();
				break;

			case MediaProfile::DVDMinusRWSeq:
			case MediaProfile::CDRW:
				Erase();
			break;
		}
	}
	else
	{
		throw BurnerException(BE_DEVICE_NOT_SET, BE_DEVICE_NOT_SET_TEXT);
	}
}

// IDeviceCallback13
void Burner::onFormatProgress(DOUBLE fPercentCompleted)
{
	if (m_pCallback)
		m_pCallback->OnFormatProgress(fPercentCompleted);
}

void Burner::onEraseProgress(DOUBLE fPercentCompleted)
{
	if (m_pCallback)
		m_pCallback->OnEraseProgress(fPercentCompleted);
}
// DataDiscCallback
void Burner::onProgress(int64_t ddwPos, int64_t ddwAll)
{
	if (m_pCallback)
		m_pCallback->OnProgress(ddwPos, ddwAll);
}

void Burner::onStatus(DataDiscStatus::Enum eStatus) 
{
	if (m_pCallback)
	{
		tstring status = GetDataDiscStatusString(eStatus);
		m_pCallback->OnStatus(status);
	}
}

void Burner::onFileStatus(int32_t fileIndex, const char_t* fileName, int32_t percentComplete) 
{
	if (m_pCallback)
	{
		tstring fileName1 = fileName;
		m_pCallback->OnFileProgress(fileIndex, fileName1, percentComplete);
	}
}

BOOL Burner::onContinueWrite() 
{
	if (m_pCallback)
		return m_pCallback->OnContinue() ? TRUE : FALSE;

	return TRUE;
}

//Protected methods
void Burner::CleanupEngine()
{
	if (NULL != m_pEngine)
	{
		m_pEngine->shutdown();
		m_pEngine->release();
	}
	m_pEngine = NULL;
}

void Burner::CleanupDeviceEnum()
{
	if (NULL != m_pEnumerator)
	{
		m_pEnumerator->release();
	}
	m_pEnumerator = NULL;
}

void Burner::CleanupDevice()
{
	if (NULL != m_pDevice)
	{
		m_pDevice->release();
	}
	m_pDevice = NULL;
}

void Burner::CleanupDataDisc()
{
	if (NULL != m_pDataDisc)
	{
		m_pDataDisc->release();
	}
	m_pDataDisc = NULL;
}

tstring Burner::get_DeviceTitle(Device* pDevice)
{
	char chLetter = pDevice->driveLetter();

	TCHAR tcsName[200] = {0};
	_stprintf(tcsName, _TEXT("(%c:) - %s"), chLetter, pDevice->description());

	return tcsName;
}

void Burner::SetPacketParameters(DataDisc* pDataDisc, const PacketBurnSettings& settings)
{
	pDataDisc->setImageType(settings.get_ImageType());

	SetVolumeProperties(pDataDisc, settings.get_VolumeLabel(), settings.get_ImageType());

	pDataDisc->setSimulateBurn(FALSE);

	// Packet mode
	pDataDisc->setWriteMethod(WriteMethod::Packet);

	// Reserve the path table
	DataWriteStrategy::Enum writeStrategy = DataWriteStrategy::None;

	switch (settings.get_Option())
	{
	case PBO_START:
		writeStrategy = DataWriteStrategy::ReserveFileTableTrack;
		break;
	case PBO_FINALIZE:
		writeStrategy = DataWriteStrategy::WriteFileTableTrack;
		break;
	}
	pDataDisc->setWriteStrategy(writeStrategy);

	pDataDisc->setCloseTrack(PBO_FINALIZE == settings.get_Option());
	pDataDisc->setCloseSession(PBO_FINALIZE == settings.get_Option());
	pDataDisc->setCloseDisc(PBO_FINALIZE == settings.get_Option());
}
int32_t Burner::GetStartAddress(Device * pDevice)
{
	assert(pDevice);

	DiscInfo *pDI = pDevice->readDiscInfo();
	if(NULL == pDI)
		return -1;

	const SessionState::Enum sessionState = pDI->sessionState();
	pDI->release();

	// Use newTrackStartAddress if the last session is open. That will give 
	// us next available address for writing in the open session.
	if (SessionState::Open == sessionState)
		return pDevice->newTrackStartAddress();

	return pDevice->newSessionStartAddress();
}

int Burner::GetLastTrackNumber(Device * pDevice, bool completeOnly)
{
	// Get the last track number from the last session if multisession option was specified
	int nLastTrack = 0;

	// Check for DVD+RW and DVD-RW RO random writable media. 
	MediaProfile::Enum mp = pDevice->mediaProfile();

	if ((MediaProfile::DVDPlusRW == mp) || (MediaProfile::DVDMinusRWRO == mp) || 
		(MediaProfile::BDRE == mp)  || (MediaProfile::BDRSrmPow == mp) || (MediaProfile::DVDRam == mp))
	{
		// DVD+RW and DVD-RW RO has only one session with one track
		if (pDevice->mediaFreeSpace() > 0)
			nLastTrack = 1;	
	}		
	else
	{
		// All other media is recorded using tracks and sessions and multi-session is no different 
		// than with the CD. 

		// Use the ReadDiskInfo method to get the last track number
		DiscInfo *pDI = pDevice->readDiscInfo();
		if(NULL != pDI)
		{
			nLastTrack = pDI->lastTrack();
			
			if (completeOnly)
			{
				// readDiskInfo reports the empty space as a track too
				// That's why we need to go back one track to get the last completed track
				if ((DiscStatus::Open == pDI->discStatus()) || (DiscStatus::Empty == pDI->discStatus()))
					nLastTrack--;
			}

			pDI->release();
		}
	}

	return nLastTrack;
}


tstring Burner::GetDataDiscStatusString(DataDiscStatus::Enum eStatus)
{
	switch(eStatus)
	{
		case DataDiscStatus::BuildingFileSystem:
			return tstring(_TEXT("Building filesystem..."));
		
		case DataDiscStatus::WritingFileSystem:
			return tstring(_TEXT("Writing filesystem..."));

		case DataDiscStatus::WritingImage:
			return tstring(_TEXT("Writing image..."));

		case DataDiscStatus::CachingSmallFiles:
			return tstring(_TEXT("Caching small files ..."));

		case DataDiscStatus::CachingNetworkFiles:
			return tstring(_TEXT("Caching network files ..."));

		case DataDiscStatus::CachingCDRomFiles:
			return tstring(_TEXT("Caching CD-ROM files ..."));

		case DataDiscStatus::Initializing:
			return tstring(_TEXT("Initializing..."));

		case DataDiscStatus::Writing:
			return tstring(_TEXT("Writing..."));

		case DataDiscStatus::WritingLeadOut:
			return tstring(_TEXT("Flushing device cache and writing lead-out..."));

		case DataDiscStatus::LoadingImageLayout:
			return tstring(_TEXT("Loading image layout from last track..."));
	}

	return tstring(_TEXT("Unknown status..."));
}

void Burner::SetVolumeProperties(DataDisc* pDataDisc, const tstring& volumeLabel, primo::burner::ImageTypeFlags::Enum imageType)
{
	// set volume times
	SYSTEMTIME st;
	GetSystemTime(&st);

	FILETIME ft;
	SystemTimeToFileTime(&st, &ft);
	
	if((ImageTypeFlags::Iso9660 & imageType) ||
		(ImageTypeFlags::Joliet & imageType))
	{
		IsoVolumeProps *iso = pDataDisc->isoVolumeProps();
		
		iso->setVolumeLabel(volumeLabel.c_str());

		// Sample settings. Replace with your own data or leave empty
		iso->setSystemID(_T("WINDOWS"));
		iso->setVolumeSet(_TEXT("SET"));
		iso->setPublisher(_T("PUBLISHER"));
		iso->setDataPreparer(_T("PREPARER"));
		iso->setApplication(_T("DVDBURNER"));
		iso->setCopyrightFile(_T("COPYRIGHT.TXT"));
		iso->setAbstractFile(_T("ABSTRACT.TXT"));
		iso->setBibliographicFile(_T("BIBLIO.TXT"));

		iso->setVolumeCreationTime(ft);
	}

	if(ImageTypeFlags::Joliet & imageType)
	{
		JolietVolumeProps *joliet = pDataDisc->jolietVolumeProps();
		
		joliet->setVolumeLabel(volumeLabel.c_str());

		// Sample settings. Replace with your own data or leave empty
		joliet->setSystemID(_T("WINDOWS"));
		joliet->setVolumeSet(_TEXT("SET"));
		joliet->setPublisher(_T("PUBLISHER"));
		joliet->setDataPreparer(_T("PREPARER"));
		joliet->setApplication(_T("DVDBURNER"));
		joliet->setCopyrightFile(_T("COPYRIGHT.TXT"));
		joliet->setAbstractFile(_T("ABSTRACT.TXT"));
		joliet->setBibliographicFile(_T("BIBLIO.TXT"));

		joliet->setVolumeCreationTime(ft);
	}

	if(ImageTypeFlags::Udf & imageType)
	{
		UdfVolumeProps *udf = pDataDisc->udfVolumeProps();
		
		udf->setVolumeLabel(volumeLabel.c_str());

		// Sample settings. Replace with your own data or leave empty
		udf->setVolumeSet(_TEXT("SET"));
		udf->setCopyrightFile(_T("COPYRIGHT.TXT"));
		udf->setAbstractFile(_T("ABSTRACT.TXT"));

		udf->setVolumeCreationTime(ft);
	}
}

File: windows/net/samples/VideoDVD/Burner.cs
using System;
using System.IO;
using System.Collections;

using PrimoSoftware.Burner;
using System.Collections.Generic;

namespace VideoDVD
{
    public class BurnerCallback
    {
        public delegate void Status(string message);
        public delegate void ImageProgress(long pos, long all);
        public delegate bool Continue();
    }

    public class ProgressInfo
    {
        public string Message = "";
        public int Percent = 0;
        public int UsedCachePercent = 0;
        public int ActualWriteSpeed = 0;
    }

    // Burn Settings
    public class BurnSettings
    {
        public string SourceFolder;
        public string VolumeLabel;
    };

    /// <summary>
    /// Container for device information
    /// </summary>
	public struct DeviceInfo
	{
        /// <summary>
        /// Device index
        /// </summary>
		public int Index;

        /// <summary>
        /// Device description
        /// </summary>
		public string Title;

        /// <summary>
        /// Returns string representation of this object
        /// </summary>
		public override string ToString() 
		{ 
			return Title; 
		}
	};


	public class Burner
	{
		public event BurnerCallback.Status Status;
		public event BurnerCallback.ImageProgress ImageProgress;
		public event BurnerCallback.Continue Continue;
		
		public bool IsOpen 
		{ 
			get 
			{ 
					return m_isOpen; 
			}	
		}

        private void CheckDevice()
        {
            if (null == m_device)
                throw new Exception(BurnerErrors.NO_DEVICE_TEXT);
        }


		public int DeviceCacheSize
		{
			get
			{
                CheckDevice();
				return m_device.InternalCacheCapacity;
			}
		}

		public int DeviceCacheUsedSpace
		{
			get
			{
                CheckDevice();
				return m_device.InternalCacheUsedSpace;
			}
		}

		public int WriteTransferKB
		{
			get
            {
                CheckDevice();
                return m_device.WriteTransferRate;
			}
		}

		public long MediaFreeSpace
		{
			get
			{
                CheckDevice();
				return m_device.MediaFreeSpace;
			}
		}

		public PrimoSoftware.Burner.MediaProfile MediaProfile
		{
			get
			{
                CheckDevice();
				return m_device.MediaProfile;
			}
		}

		public string MediaProfileString
		{
			get
			{
				PrimoSoftware.Burner.MediaProfile profile = this.MediaProfile;
				switch(profile)
				{
					case MediaProfile.CdRom:
						return "CD-ROM";

					case MediaProfile.CdR:
						return "CD-R";

					case MediaProfile.CdRw:
						return "CD-RW";

					case MediaProfile.DvdRom:
						return "DVD-ROM";

					case MediaProfile.DvdMinusRSeq:
						return "DVD-R Sequential Recording.";

					case MediaProfile.DvdMinusRDLSeq:
						return "DVD-R DL 8.54GB for Sequential Recording.";

					case MediaProfile.DvdMinusRDLJump:
						return "DVD-R DL 8.54GB for Layer Jump Recording.";

					case MediaProfile.DvdRam:
						return "DVD-RAM";

					case MediaProfile.DvdMinusRwRo:
						return "DVD-RW Restricted Overwrite.";

					case MediaProfile.DvdMinusRwSeq:
						return "DVD-RW Sequential Recording.";

					case MediaProfile.DvdPlusRw:
					{
						BgFormatStatus fmt = m_device.BgFormatStatus;
						switch(fmt)
						{
							case BgFormatStatus.NotFormatted:
								return "DVD+RW. Not formatted.";
							case BgFormatStatus.Partial:
								return "DVD+RW. Partially formatted.";
							case BgFormatStatus.Pending:
								return "DVD+RW. Background formatting is pending ...";
							case BgFormatStatus.Completed:
								return "DVD+RW. Formatted.";
						}
						return "DVD+RW.";
					}

					case MediaProfile.DvdPlusR:
						return "DVD+R";

					case MediaProfile.DvdPlusRDL:
						return "DVD+R DL 8.5GB";

					default:
						return "Unknown Profile.";
				}
			}
		}


		#region Public Methods
		public void Open()
		{	
			if (m_isOpen)
				return;

			// Enable trace log
			Library.EnableTraceLog(null, true);

			m_engine = new Engine();
			if (!m_engine.Initialize()) 
			{
				m_engine.Dispose();
				m_engine = null;

                throw new BurnerException(m_engine.Error);
			}

			m_isOpen = true;
		}

		public void Close()
		{
			if (null != m_device)
				m_device.Dispose();
			m_device = null;

			if (null != m_engine)
			{
				m_engine.Shutdown();
				m_engine.Dispose();
			}
			m_engine = null;

			Library.DisableTraceLog();

			m_isOpen = false;
		}

		public List<DeviceInfo> EnumerateDevices()
		{
            if (!m_isOpen)
                return null;

			m_devices.Clear();

            using (DeviceEnumerator enumerator = m_engine.CreateDeviceEnumerator())
            {
                int devices = enumerator.Count;

                for (int i = 0; i < devices; i++)
                {
                    Device device = enumerator.CreateDevice(i);
                    if (null != device)
                    {
                        DeviceInfo dev = new DeviceInfo();
                        dev.Index = i;
                        dev.Title = GetDeviceTitle(device);

                        m_devices.Add(dev);

                        device.Dispose();
                    }
                }
            }

			return m_devices;
		}

		public bool SelectDevice(int deviceIndex, bool exclusive)
		{
            if (m_device != null)
                return false;

            using (DeviceEnumerator enumerator = m_engine.CreateDeviceEnumerator())
            {
                Device dev = enumerator.CreateDevice(deviceIndex, exclusive);
                if (null == dev)
                    throw new BurnerException(enumerator.Error);

                m_device = dev;
                
            }

            return true;
		}

		public void ReleaseDevice()
		{
			if (null != m_device)
				m_device.Dispose();

			m_device = null;
		}


		public long CalculateImageSize(string sourceFolder)
		{
            using (DataDisc data = new DataDisc())
            {
                data.DvdVideo = true;
                SetImageLayoutFromFolder(data, sourceFolder);
                return data.ImageSizeInBytes;
            }
		}

		public void Burn(BurnSettings settings)
        {
            if (null == m_device)
                return;

            using (DataDisc data = new DataDisc())
            {
                // Add event handlers
                data.OnStatus += new EventHandler<DataDiscStatusEventArgs>(DataDisc_OnStatus);
                data.OnProgress += new EventHandler<DataDiscProgressEventArgs>(DataDisc_OnProgress);
                data.OnContinueBurn += new EventHandler<DataDiscContinueEventArgs>(DataDisc_OnContinueBurn);

                // Format if DVD+RW
                FormatMedia(m_device);

                m_device.WriteSpeedKB = m_device.MaxWriteSpeedKB;

                data.Device = m_device;
                data.WriteMethod = WriteMethod.DvdDao;
                data.CloseDisc = true;

                // Set the session start address. This must be done before intializing the file system.
                data.SessionStartAddress = m_device.NewSessionStartAddress;

                data.DvdVideo = true;

                SetVolumeProperties(data, settings.VolumeLabel);

                SetImageLayoutFromFolder(data, settings.SourceFolder);

                // Burn. Dismount after burning is complete
                if (!data.WriteToDisc(true))
                    throw new BurnerException(data.Error);

                m_device.Eject(true);
            }

        }

		#endregion
		
		#region DataDisc Event Handlers
		public void DataDisc_OnStatus(Object sender, DataDiscStatusEventArgs args)
		{
			if (null == Status)
				return;

			Status(GetDataDiscStatusString(args.Status));
		}
		

		public void DataDisc_OnProgress(Object sender, DataDiscProgressEventArgs args)
		{
			if (null == ImageProgress)
				return;

			ImageProgress(args.Position, args.All);
		}

		public void DataDisc_OnContinueBurn(Object sender, DataDiscContinueEventArgs args)
		{
            if (null == Continue)
            {
                args.Continue = true;
            }
			
            args.Continue = Continue();
		}
		#endregion

		#region Private Methods
			private string GetDeviceTitle(Device device)
			{
				return String.Format("({0}:) - {1}", device.DriveLetter, device.Description);
			}


			private void SetImageLayoutFromFolder(DataDisc data, string sourceFolder)
            {
                DataFile dvdFileSystemRoot = null;
                using (PrimoSoftware.Burner.VideoDVD dvd = new PrimoSoftware.Burner.VideoDVD())
                {
                    // pass the source folder, the layout will be created by PrimoBurner
                    if (!dvd.SetImageLayoutFromFolder(sourceFolder))
                    {
                        throw new BurnerException(dvd.Error);
                    }

                    // Get the correct dvd layout 
                    dvdFileSystemRoot = dvd.ImageLayout;
                    if (!data.SetImageLayout(dvdFileSystemRoot))
                    {
                        throw new BurnerException(data.Error);
                    }
                }
            }

			private void SetVolumeProperties(DataDisc data, string volumeLabel)
			{
                DateTime creationTime = DateTime.Now;

                // DVD-Video uses UDF and ISO filesystems
				data.IsoVolumeProps.VolumeLabel = volumeLabel;
                data.IsoVolumeProps.CreationTime = creationTime;
                data.UdfVolumeProps.VolumeLabel = volumeLabel;
                data.UdfVolumeProps.CreationTime = creationTime;
			}

			private string GetDataDiscStatusString(DataDiscStatus status)
			{
				switch(status)
				{
					case DataDiscStatus.BuildingFileSystem:
						return "Building filesystem...";
					case DataDiscStatus.LoadingImageLayout:
						return "Loading image layout...";
					case DataDiscStatus.WritingFileSystem:
						return "Writing filesystem...";
					case DataDiscStatus.WritingImage:
						return "Writing image...";
					case DataDiscStatus.CachingSmallFiles:
						return "Caching small files...";
					case DataDiscStatus.CachingNetworkFiles:
						return "Caching network files...";
					case DataDiscStatus.CachingCDRomFiles:
						return "Caching CDROM files...";
					case DataDiscStatus.Initializing:
						return "Initializing and writing lead-in...";
					case DataDiscStatus.Writing:
						return "Writing...";
					case DataDiscStatus.WritingLeadOut:
						return "Writing lead-out and flushing cache...";
				}

				return "Unknown status...";
			}

			void FormatMedia(Device dev)
			{
				switch(dev.MediaProfile)
				{
					// DVD+RW (needs to be formatted before the disc can be used)
					case MediaProfile.DvdPlusRw:
					{
						if (null != Status)
							Status("Formatting...");

						switch(dev.BgFormatStatus)
						{
							case BgFormatStatus.NotFormatted:
								dev.Format(FormatType.DvdPlusRwFull);
							break;

							case BgFormatStatus.Partial:
								dev.Format(FormatType.DvdPlusRwRestart);
							break;
						}
					}
					break;
				}
			}

		#endregion
	

			private bool m_isOpen = false;
            private List<DeviceInfo> m_devices = new List<DeviceInfo>();

			private Engine m_engine = null;
			private Device m_device = null;
	}
}
